R/pkg/R/functions.R

```diff
@@ -2230,6 +2237,32 @@ setMethod("from_json", signature(x = "Column", schema = "characterOrstructType")
             column(jc)
           })
 
+#' @details
+#' \code{schema_of_json}: Parses a JSON string and infers its schema in DDL format.
+#'
+#' @rdname column_collection_functions
+#' @aliases schema_of_json schema_of_json,characterOrColumn-method
+#' @examples
+#'
+#' \dontrun{
+#' json <- '{"name":"Bob"}'
+#' df <- sql("SELECT * FROM range(1)")
+#' head(select(df, schema_of_json(json)))}
+#' @note schema_of_json since 3.0.0
+setMethod("schema_of_json", signature(x = "characterOrColumn"),
+          function(x, ...) {
+            if (class(x) == "character") {
+              col <- callJStatic("org.apache.spark.sql.functions", "lit", x)
+            } else {
+              col <- x@jc
```

Comment:
what's the use when x is a Column?
`schema_of_csv(lit("Amsterdam,2018")))` seems a bit odd to me...

Reply:
That's actually related with Scala API. There are too many overridden versions of functions in `function.scala` so we're trying to reduce it. Column is preferred over other specific types because Column can cover other expression cases.. in Python or R, they can be easily supported so other types and column are all supported. To cut it short, for consistency with Scala API.

Reply:
ok but one use could be 

`select(df, schema_of_csv(df$schemaCol))`

like an actual col not a literal string?

Reply:
Yea .. that was discussed at https://github.com/apache/spark/pull/22775. The usecase of `schema_of_csv` or `schema_of_json` will usually be like .. copying and pasting one record from the actual data manually. That's disallowed for now conservatively.

Reply:
you are saying this `select(df, schema_of_csv(df$schemaCol))` is not allowed?

Reply:
Yup.. only literal works but columns don't work.

Reply:
hm.. why not just support string then? it's kinda very odd usage in R `schema_of_csv(lit("Amsterdam,2018")))`

Reply:
It's just left as column for now .. in case we allow other more cases .. Yea, it's a bit odd that we should `schema_of_csv(lit("Amsterdam,2018")))`. Maybe later columns or other expressions might have to be supported - in that case we can just fix the documentation in R side ..

Reply:
BTW, `lit` usage already works in many APIs although it looks a bit odd .. should be okay.

Reply:
just that I thought the shortcut syntax in scala is nicer looking then `lit("string")` in R

Reply:
Hmm .. do you mind if we go ahead for this one and talk later within 3.0? I think we're going to deal with this problem within 3.0 if I am not mistaken. I need to make one followup after this anyway.

Reply:
maybe to think about the design of API in R and Scala and else where - what does it look like when the user passes in a column that is not a literal string? probably worthwhile to follow up separately.